keps/sig-release/20190121-artifact-management.md

```diff
@@ -109,6 +109,28 @@ Importantly, if a project needs to roll-back or remove an artifact, the same pro
 occur, so that the promotion tool needs to be capable of deleting images and artifacts as
 well as promoting them.
 
+### HTTP Redirector Design
+To facilitate world-wide distribution of artifacts from a single (virtual) location we will
+ideally run a replicated redirector service in the United States, Europe and Asia.
+Each of these redirectors
+services will be deployed in a Kubernetes cluster and they will be exposed via a public IP
+address and a dns record indicating their location (e.g. `europe.artifacts.k8s.io`).
+
+We will use Geo DNS to route requests to `artifacts.k8s.io` to the correct redirector. This is necessary to ensure that we always route to a server which is accessible no matter what region we are in. We will need to extend or enhance the existing DNS synchronization tooling to handle creation of the GeoDNS records.
```

Comment:
GCP (our current DNS) doesn't currently support GeoDNS.  If Geo is the only solution, we may have to overhaul that.  Before calling for geo-dns as a solution, though, we should make sure we capture the requirements.

It sounds like you're aiming for:

1) a single DNS name for end-users (rather than pushing geo onto them)
2) support for mirrors other than Google Cloud (which has access problems in China)
3) support for per-cloud mirrors

Are there more?

#1 can be addressed via anycast.  If not for #2 and #3, that might be viable.

#2 might be mitigatable within the existing hosting.  I have ideas on how to set that up, but it would need some testing, so we would need help from someone who actually suffers from the China blockages.

#3 is the real difficult one.  It's not clear to me how to handle this - geodns is not quite the right  model, since most clouds share most geographies, but it's still better to serve an aws user from within aws.

Lastly, I want to keep in mind that this all comes at some cost.  Someone will have to foot the bills for this.  Let's not downplay that hurdle.  We should shoot for a model that is extensible to arbitrary mirrors, if possible.

Do any of the existing CDN products out there help us?  Rather than build a CDN of our own...

Reply:
@thockin 

I don't actually think this is that hard, it's not really a CDN, we're not going to host the blobs (we'll use a real CDN for that) but we are going to host the redirector.

When I said "GeoDNS" it means more than just "auto-geo"

Most clouds publish their IP ranges, e.g. [AWS](https://docs.aws.amazon.com/general/latest/gr/aws-ip-ranges.html),  [Azure](https://www.microsoft.com/en-us/download/details.aspx?id=56519) and [GCP](https://cloud.google.com/compute/docs/faq#find_ip_range)

So I think we can definitely determine which cloud a user is coming from and route accordingly.

Reply:
To further clarify:

GeoDNS gets us to the redirector. Redirector makes a decision based on cloud, and then Geo information.

Reply:
Yeah, I conflated a bit there.  Sorry.

What we don't have here is requirements for the redirector itself.  If this is to be in the path of every `docker pull` for k8s resources, it has to be DAMNED reliable.

One of the nice things about GCR/GCS (or s3 or ...) is that this becomes someone else's problem.  Those services are understood to be pretty reliable, overall.

So we need a globally available redirector service with very high reliability and low latency.  That redirector needs access to a federation of backend mirrors and rules to map incoming traffic to lists of viable mirrors.

We need some way to prove image veracity and non-tamperedness.

We need to treat this like a real production service.  That means monitoring and on-calls and SLOs.

We probably need to health-check backend mirrors.

We need a notification and sync mechanism so backends can know when and what to mirror (e.g. rsync doesn't work against S3).

I don't know that we need GeoDNS to get to the redirector.  We can stay within the existing infrastructure and use GCP's anycast capabilities and multi-cluster Ingress.  it's my understanding that there's no problem with any of that from mainland China -- only things like googleapis.com.   If we need to expand beyond that, we may need GeoDNS or something, yeah.